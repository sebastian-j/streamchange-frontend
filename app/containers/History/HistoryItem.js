import React from 'react';
import PropTypes from 'prop-types';
import RelativeDate from '../../components/RelativeDate';

const HistoryItem = props => (
  <tr className="md-tr">
    <td className="md-td md-td-image">
      <a
        href={`https://www.youtube.com/channel/${props.channelId}`}
        target="_blank"
      >
        <img alt="Logo" src={props.imageUrl} height="45px" />
      </a>
    </td>
    <td className="md-td md-td-text-left">
      <span>{props.displayName}</span>
    </td>
    <td className="md-td md-td-text">
      <span>{props.prize}</span>
    </td>
    <td className="md-td md-td-text">
      <span>{props.message}</span>
    </td>
    <td className="md-td md-td-text">
      <RelativeDate ISO8601Date={props.createdAt} />
    </td>
  </tr>
);

HistoryItem.propTypes = {
  channelId: PropTypes.string.isRequired,
  imageUrl: PropTypes.string,
  displayName: PropTypes.string,
  prize: PropTypes.string,
  message: PropTypes.number,
  createdAt: PropTypes.string,
};
export default HistoryItem;
