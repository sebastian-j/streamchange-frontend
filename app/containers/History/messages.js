/*
 * History Messages
 *
 * This contains all the text for the History container.
 */
import { defineMessages } from 'react-intl';

export const scope = 'streamchange.containers.History';

export default defineMessages({
  dateHeader: {
    id: `${scope}.header.date`,
    defaultMessage: 'Date and time',
  },
  messageHeader: {
    id: `${scope}.header.message`,
    defaultMessage: 'Message',
  },
  nameHeader: {
    id: `${scope}.header.name`,
    defaultMessage: 'Channel name',
  },
  prizeHeader: {
    id: `${scope}.header.prize`,
    defaultMessage: 'Prize',
  },
  infoError: {
    id: `${scope}.info.error`,
    defaultMessage: 'Cannot load list',
  },
  infoLoading: {
    id: `${scope}.info.loading`,
    defaultMessage: 'Loading history',
  },
  infoNoResults: {
    id: `${scope}.info.noResults`,
    defaultMessage: 'No channel match the search term',
  },
  resultsPerPage: {
    id: `${scope}.resultsPerPage`,
    defaultMessage: 'Results per page',
  },
  returnButton: {
    id: `${scope}.ReturnButton`,
    defaultMessage: 'Back to giveaway',
  },
  searchLabel: {
    id: `${scope}.search.label`,
    defaultMessage: 'Search',
  },
  clearHistoryMenuItem: {
    id: `${scope}.HistoryMenu.clearHistory`,
    defineMessage: 'Clear history',
  },
  clearHistoryCancelButton: {
    id: `${scope}.HistoryMenu.clearHistory.cancelButton`,
    defineMessage: 'Cancel',
  },
  clearHistoryConfirmButton: {
    id: `${scope}.HistoryMenu.clearHistory.confirmButton`,
    defineMessage: 'OK',
  },
  clearHistoryDialogContent: {
    id: `${scope}.HistoryMenu.clearHistory.dialogContentText`,
    defineMessage: 'Are you sure you want to clear all winner history?',
  },
});
