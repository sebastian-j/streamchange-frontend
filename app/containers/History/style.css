.page-wrapper {
    padding: 10px;
    overflow-x: hidden;
    width: 100%;
}

.header-btn {
    background: none;
    border: none;
    padding: 14px 10px 14px 16px;
    outline: none;
    width: 100%;
}

.active {
    color: black;
    opacity: 1;
}

.icon-direction-desc {
    transform: rotate(180deg);
}

.sort-icon:not(.active) {
    opacity: 0;
}

.sort-icon {
    transition: opacity 300ms ease-out 0ms, transform 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms !important;
}

.md-table {
    display: table;
    overflow-y: auto;
    min-width: 1052px;
    width: 100%;
}

.md-thead {
    border-bottom: 1px solid rgba(224, 224, 224, 1);
    cursor: pointer;
    user-select: none;
}

.md-tr {
    height: 50px;
}

.md-td {
    border-bottom: 1px solid rgba(224, 224, 224, 1);
    border-spacing: 0;
    border-collapse: collapse;
    font-family: "Roboto", "Helvetica", "Arial", sans-serif;
}

.md-td-text {
    font-size: 0.875rem;
    padding: 14px 30px 14px 16px;
    text-align: right;
}

.md-td-text-left {
    font-size: 1rem;
    padding: 14px 30px 14px 16px;
    text-align: left;
}

.md-td-image {
    max-width: 100px;
}

.md-footer {
    display: flex;
    flex-direction: row-reverse;
}

@media (orientation: portrait) {
    .page-wrapper {
        overflow-x: auto !important;
    }
}