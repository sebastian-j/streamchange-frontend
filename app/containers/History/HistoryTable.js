import React, { useState } from 'react';
import PropTypes from 'prop-types';
import clsx from 'clsx';
import styled from 'styled-components';
import ArrowUpIcon from './arrowUpIcon';
import HistoryItem from './HistoryItem';
import './style.css';

const HeaderButton = styled.button`
  background: none;
  border: none;
  color: ${props => props.theme.staticTextColor};
  padding: 14px 10px 14px 16px;
  outline: none;
  width: 100%;
`;

const Table = styled.table`
  display: table;
  overflow-y: auto;
  min-width: 1052px;
  width: 100%;
`;

const Thead = styled.thead`
  border-bottom: 1px solid rgba(224, 224, 224, 1);
  cursor: pointer;
  user-select: none;
`;

const HistoryTable = props => {
  const [sort, setSort] = useState('createdAtDESC');
  const handleSortChange = event => {
    const { target } = event;
    let value = target.id;
    if (value === sort) {
      value += 'DESC';
    }
    setSort(value);
  };

  return (
    <Table>
      <Thead>
        <tr>
          <td />
          <td>
            <HeaderButton
              id="displayName"
              onClick={handleSortChange}
              type="button"
            >
              Nazwa
              <ArrowUpIcon
                className={clsx(
                  'sort-icon',
                  sort === 'displayName' && 'active',
                  sort === 'displayNameDESC' && [
                    'active',
                    'icon-direction-desc',
                  ],
                )}
              />
            </HeaderButton>
          </td>
          <td>
            <HeaderButton id="prize" onClick={handleSortChange} type="button">
              Nagroda
              <ArrowUpIcon
                className={clsx(
                  'sort-icon',
                  sort === 'prize' && 'active',
                  sort === 'prizeDESC' && ['active', 'icon-direction-desc'],
                )}
              />
            </HeaderButton>
          </td>
          <td>
            <HeaderButton id="message" onClick={handleSortChange} type="button">
              Wiadomość
              <ArrowUpIcon
                className={clsx(
                  'sort-icon',
                  sort === 'message' && 'active',
                  sort === 'messageDESC' && ['active', 'icon-direction-desc'],
                )}
              />
            </HeaderButton>
          </td>
          <td>
            <HeaderButton
              id="createdAt"
              onClick={handleSortChange}
              type="button"
            >
              Data i godzina
              <ArrowUpIcon
                className={clsx(
                  'sort-icon',
                  sort === 'createdAt' && 'active',
                  sort === 'createdAtDESC' && ['active', 'icon-direction-desc'],
                )}
              />
            </HeaderButton>
          </td>
        </tr>
      </Thead>
      <tbody>
        {props.items.map(item => (
          <HistoryItem
            channelId={item.channelId}
            imageUrl={item.imageUrl}
            displayName={item.displayName}
            prize={item.prize}
            message={item.message}
            createdAt={item.createdAt}
          />
        ))}
      </tbody>
    </Table>
  );
};

HistoryTable.propTypes = {
  items: PropTypes.array,
};

export default HistoryTable;
