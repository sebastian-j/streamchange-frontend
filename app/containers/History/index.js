import React from 'react';
import { NavLink } from 'react-router-dom';
import styled from 'styled-components';
import LinearProgress from '@material-ui/core/LinearProgress';
import FormControl from '@material-ui/core/FormControl';
import InputLabel from '@material-ui/core/InputLabel';
import MenuItem from '@material-ui/core/MenuItem';
import Select from '@material-ui/core/Select';
import IconButton from '@material-ui/core/IconButton';
import db from '../../components/YoutubeWorker/db';
import StyledTextField from '../../components/StyledTextField';
import HistoryTable from './HistoryTable';

const PageWrapper = styled.div`
  background-color: ${props => props.theme.bodyBackground};
  padding: 10px;
  overflow-x: hidden;
  min-height: 100vh;
  width: 100%;
`;

const ReturnButton = styled.div`
  color: ${props => props.theme.staticTextColor};
  line-height: 48px;
  min-height: 48px;
  & span {
    margin-left: 24px;
    text-decoration: none;
  }
  &:hover {
    background: rgba(0, 0, 0, 0.2);
  }
`;

const StyledFormControl = styled(FormControl)`
  width: 128px;
  div.MuiInput-input {
    color: ${props => props.theme.staticTextColor};
  }
  label {
    color: ${props => props.theme.inputLabel};
  }
  label.Mui-focused {
    color: ${props => props.theme.inputLabelFocused};
  }
  svg.MuiSelect-icon {
    color: ${props => props.theme.secondaryTextColor};
  }
  .MuiInput-underline:before {
    border-bottom-color: ${props => props.theme.secondaryTextColor};
  }
`;

const Information = styled.div`
  color: ${props => props.theme.staticTextColor};
  font-size: 2vw;
  margin-top: 5vh;
  text-align: center;
`;
const Footer = styled.div`
  display: flex;
  flex-direction: row-reverse;
  margin-top: 10px;
`;

export default class History extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      error: false,
      isLoaded: false,
      items: [],
      search: '',
      maxResults: 20,
      page: 0,
      isLastPage: false,
    };
    this.handleChange = this.handleChange.bind(this);
    this.prevPage = this.prevPage.bind(this);
    this.nextPage = this.nextPage.bind(this);
    this.getHistory = this.getHistory.bind(this);
  }

  handleChange(event) {
    const { target } = event;
    const { value } = target;
    const { name } = target;
    this.setState(
      {
        [name]: value,
      },
      () => {
        this.getHistory();
      },
    );
  }

  prevPage() {
    const oldPage = this.state.page;
    if (this.state.page > 0) {
      this.setState({ page: oldPage - 1 }, () => {
        this.getHistory();
      });
    }
  }

  nextPage() {
    const oldPage = this.state.page;
    this.setState({ page: Number(oldPage + 1) }, () => {
      this.getHistory();
    });
  }

  getHistory() {
    const firstResult = Number(this.state.page * this.state.maxResults);
    db.table('history')
      .filter(winner =>
        winner.displayName
          .toLowerCase()
          .includes(this.state.search.toLowerCase()),
      )
      .reverse()
      .offset(firstResult)
      .limit(this.state.maxResults)
      .toArray()
      .then(items => {
        this.setState({ isLoaded: true, error: false, items });
        if (this.state.maxResults > items.length) {
          this.setState({ isLastPage: true });
        } else this.setState({ isLastPage: false });
      })
      .catch(() => {
        this.setState({
          isLoaded: true,
          error: true,
        });
      });
  }

  componentDidMount() {
    this.getHistory();
  }

  render() {
    if (this.state.error) {
      return <div>Nie udało się załadować listy.</div>;
    }
    if (!this.state.isLoaded) {
      return (
        <PageWrapper>
          <LinearProgress />
          <div
            style={{
              fontSize: '2vw',
              position: 'absolute',
              left: '55%',
              top: '50%',
              transform: 'translate(-50%, -50%)',
              textAlign: 'center',
            }}
          >
            Ładowanie listy
          </div>
        </PageWrapper>
      );
    }
    return (
      <PageWrapper>
        <NavLink to="/giveaway" activeClassName="active">
          <ReturnButton>
            <svg
              xmlns="http://www.w3.org/2000/svg"
              width="24"
              height="24"
              viewBox="0 0 24 24"
            >
              <path d="M0 0h24v24H0z" fill="none" />
              <path d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z" />
            </svg>
            <span>Powrót do losowania</span>
          </ReturnButton>
        </NavLink>
        <StyledTextField
          id="search"
          name="search"
          label="Wyszukaj"
          value={this.state.search}
          onChange={this.handleChange}
          type="text"
          margin="normal"
          fullWidth
        />
        {this.state.search.length > 0 && this.state.items.length === 0 && (
          <Information>
            Żaden kanał nie pasuje do wyszukiwanego słowa.
          </Information>
        )}
        {this.state.items.length > 0 && (
          <HistoryTable items={this.state.items} />
        )}
        <Footer>
          {!this.state.isLastPage && (
            <IconButton
              edge="end"
              aria-label="Next page"
              name="next"
              onClick={this.nextPage}
            >
              <svg
                xmlns="http://www.w3.org/2000/svg"
                width="24"
                height="24"
                viewBox="0 0 24 24"
              >
                <path d="M8.59 16.59L13.17 12 8.59 7.41 10 6l6 6-6 6-1.41-1.41z" />
                <path fill="none" d="M0 0h24v24H0V0z" />
              </svg>
            </IconButton>
          )}
          {this.state.page !== 0 && (
            <IconButton
              edge="end"
              aria-label="Previous page"
              name="prev"
              onClick={this.prevPage}
            >
              <svg
                xmlns="http://www.w3.org/2000/svg"
                width="24"
                height="24"
                viewBox="0 0 24 24"
              >
                <path d="M15.41 16.59L10.83 12l4.58-4.59L14 6l-6 6 6 6 1.41-1.41z" />
                <path fill="none" d="M0 0h24v24H0V0z" />
              </svg>
            </IconButton>
          )}
          <StyledFormControl>
            <InputLabel htmlFor="age-simple">Ilość na stronę</InputLabel>
            <Select
              value={this.state.maxResults}
              onChange={this.handleChange}
              inputProps={{
                name: 'maxResults',
                id: 'maxResults',
              }}
            >
              <MenuItem value={10}>10</MenuItem>
              <MenuItem value={20}>20</MenuItem>
              <MenuItem value={50}>50</MenuItem>
              <MenuItem value={100}>100</MenuItem>
            </Select>
          </StyledFormControl>
        </Footer>
      </PageWrapper>
    );
  }
}
