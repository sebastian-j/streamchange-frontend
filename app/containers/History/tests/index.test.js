import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';

import History from '../index';

const renderer = new ShallowRenderer();

describe('<History />', () => {
  it('should render and match the snapshot', () => {
    renderer.render(<History />);
    const renderedOutput = renderer.getRenderOutput();
    expect(renderedOutput).toMatchSnapshot();
  });
});
