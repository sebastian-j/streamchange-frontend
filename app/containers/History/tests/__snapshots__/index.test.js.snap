// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<History /> should render and match the snapshot 1`] = `
<ForwardRef>
  <ForwardRef(WithStyles) />
  <div
    style={
      Object {
        "fontSize": "2vw",
        "left": "55%",
        "position": "absolute",
        "textAlign": "center",
        "top": "50%",
        "transform": "translate(-50%, -50%)",
      }
    }
  >
    ≈Åadowanie listy
  </div>
</ForwardRef>
`;
